src/cloudConsole/cloudConsole.ts

```diff
@@ -343,36 +343,54 @@ export function createCloudConsole(subscriptionProvider: AzureSubscriptionProvid
             liveServerQueue = serverQueue;
 
             const tenants = await subscriptionProvider.getTenants();
-            serverQueue.push({ type: 'log', args: [localize('foundTenants', `Found ${tenants.length} tenant${tenants.length > 1 ? 's' : ''}.`)] });
             let selectedTenant: TenantIdDescription | undefined = undefined;
-            // handle multi tenant scenario, user must pick a tenant
-            if (tenants.length > 1) {
-                serverQueue.push({ type: 'log', args: [localize('selectingTenant', `Selecting tenant...`)] });
-                const picks = tenants.map(tenant => {
-                    const defaultDomainName: string | undefined = tenant.defaultDomain;
-                    return <IAzureQuickPickItem<TenantIdDescription>>{
-                        label: tenant.displayName,
-                        description: defaultDomainName,
-                        data: tenant,
-                    };
-                }).sort((a, b) => a.label.localeCompare(b.label));
-
-                const pick = await window.showQuickPick<IAzureQuickPickItem<TenantIdDescription>>(picks, {
-                    placeHolder: localize('selectDirectoryPlaceholder', "Select tenant"),
-                    ignoreFocusOut: true // The terminal opens concurrently and can steal focus (https://github.com/microsoft/vscode-azure-account/issues/77).
+
+            if (tenants.length <= 1) {
+                serverQueue.push({ type: 'log', args: [localize('foundOneTenant', `Found 1 tenant.`)] });
+                // if they have only one tenant, use it
+                selectedTenant = tenants[0];
+            } else {
+                // If the user has multiple tenants, then we check which tenants have subscriptions.
+                // This also checks if this tenant is authenticated.
+                // If a tenant is not authenticated, users will have to use the "Sign in to Directory..." command before launching cloud shell.
+                const tenantsIdsWithSubs = new Set<string>();
+                const subscriptions = await subscriptionProvider.getSubscriptions(false);
+                subscriptions.forEach((sub) => {
+                    tenantsIdsWithSubs.add(sub.tenantId);
                 });
+                const tenantsWithSubs = tenants.filter(tenant => tenantsIdsWithSubs.has(nonNullProp(tenant, 'tenantId')));
```

Comment:
Couldn't you just create an array with the set instead of filtering the tenants array?

`const tenantsWithSubs = Array.from(tenantsIdsWithSubs);`